### PR TITLE
Re-organizing lexer struct

### PR DESCRIPTION
**Pull request by bonjune**
I slightly made some changes to the lexer for better organization of the struct

- Changed `Vec<IndentationLevel>` into a new struct `Indentations` and limited its methods just for `pop`, `push`, `is_empty`, and `current`: Checking empty semantics is different from normal vectors in that `is_empty()` should return when the length is one, rather than zero because of the default indentation level.
- Made a struct for `chr0, chr1, chr2`, which is `CharWindow<const N: usize>`
- Changed condition expression in if statements. For example `if x == Some(1) || x == Some(2) { ... }` changed into `if matches!(x, Some(1 | 2))`
- Shortcircuited an if statement to reduce indentation level.
- Add a new test for `123_45_67_890` which was missing in `lexer.rs` tests module

### CODE REVIEW

File: compiler/parser/src/lexer.rs [hunk 1]
```diff
@@ -56,15 +71,89 @@ impl IndentationLevel {
     }
 }
 
+#[derive(Debug)]
+struct Indentations {
+    indent_stack: Vec<IndentationLevel>,
+}
+
+impl Indentations {
+    pub fn is_empty(&self) -> bool {
```
Comment (DimitrisJim) [inline compiler/parser/src/lexer.rs:80]: don't think there's much point in pub functions on a private struct.

File: compiler/parser/src/lexer.rs [hunk 2]
```diff
@@ -56,15 +71,89 @@ impl IndentationLevel {
...
+
+impl Default for Indentations {
+    fn default() -> Self {
+        Self {
+            indent_stack: vec![IndentationLevel::default()],
+        }
+    }
+}
+
+struct CharWindow<const N: usize>(pub [Option<char>; N]);
```
Comment (DimitrisJim) [inline compiler/parser/src/lexer.rs:110]: similarly for this `pub` here.

Comment (bonjune) [reply compiler/parser/src/lexer.rs]: Thx, I've changed those you pointed out!

File: compiler/parser/src/lexer.rs [hunk 3]
```diff
@@ -56,15 +71,89 @@ impl IndentationLevel {
...
+}
+
+impl<const N: usize, Idx> IndexMut<Idx> for CharWindow<N>
+where
+    Idx: SliceIndex<[Option<char>], Output = Option<char>>,
+{
+    fn index_mut(&mut self, index: Idx) -> &mut Self::Output {
+        self.0.index_mut(index)
+    }
+}
```
Comment (youknowone) [inline compiler/parser/src/lexer.rs:146]:
```suggestion
impl<const N: usize> Default for CharWindow<N> {
    fn default() -> Self {
        Self([None; N])
    }
}
```
We will not need `IndexMut` by adding a proper method to CharWindow

File: compiler/parser/src/lexer.rs [hunk 4]
```diff
@@ -11,10 +11,25 @@ use num_traits::identities::Zero;
...
+    make_tokenizer_located(source, Location::new(0, 0))
+}
+
+pub fn make_tokenizer_located(
+    source: &str,
+    start_location: Location,
+) -> impl Iterator<Item = LexResult> + '_ {
+    let nlh = NewlineHandler::new(source.chars());
+    Lexer::new(nlh, start_location)
+}
```
Comment (youknowone) [inline compiler/parser/src/lexer.rs:31]: Could you keep the position of these functions to keep the diff smaller?

**Review by youknowone — CHANGES_REQUESTED** [review]
Thank you for contributing! I left a few comments.

**Comment by bonjune** [comment]
Removing `IndexMut`, I added a `CharWindow::swap_first` because we need mutation for collapsing EOL characters.

**Comment by bonjune** [comment]
I think there is still room for enhancement.

Whenever a struct uses `CharWindow`, it should implement its working mechanism with `input`(in `Lexer` struct) or `source` (in `NewlineHandler` struct), such as
* initializing the CharWindow and
* what value for `CharWindow` to slide with.

I think those mechanisms should be provided by `CharWindow` wholly. It means `CharWindow` should own `T: impl Iterator<Item = char>`.

**Comment by bonjune** [comment]
I've made the changes I mentioned above. Sorry for requesting extra reviews.

**Review by youknowone — APPROVED** [review]
looks good, thank you!

**Comment by youknowone** [comment]
oh, it has conflict. could you rebase it? 

@charliermarsh do you have any advice about this?

**Comment by fanninpm** [comment]
@bonjune Could you please rebase this branch on a fresh copy of `main`?

**Comment by charliermarsh** [comment]
I haven't reviewed this thoroughly, but I'm guessing the conflict is from #4266. The main changes therein were to ensure that we emit `Tok::Comment` when lexing comments.

**Comment by bonjune** [comment]
I've just rebased. Works fine on my machine!

**Comment by youknowone** [comment]
@charliermarsh oh, i'd like to ask you code review, but that was also a good feedback, thank you!

**Pull request merged**